rfcs/20200424-sparse-domain-isolation.md

```diff
@@ -0,0 +1,649 @@
+| **Updated**   | 2020-05-23                                           |
+
+## Background
+
+In recent years, many deep learning components (such as FC, RNN, etc.) and online-learning policy were introduced to the recommender systems that result in the special framework for recommender system (such as DiFacto, etc.) gradually unable to meet the demand. More and more algorithm engineers turn to use open-source general frameworks(such as Tensorflow, etc.)  to improve production efficiency.
+In the industrial recommendation scenarios, the parameters are high-dimensional sparse with the following characteristics:
+1. The theoretical total number of dynamic features is 2^64, the size of sparse weights is more than 100GB or even several TB which make reserving memory at the beginning too expensive.
+2. The online-learning tracks user interests in real time and requires eliminating and adding weights on running-time without frequently restart training.
+
+![Models Weights Compare](20200424-sparse-domain-isolation/models-weights-compare.png)
+
+## Objective
+
+TensorFlow supports large-scale dynamic sparse weights training.
+
+## Problem & Motivation
+
+TensorFlow is well designed, but not support creating and training the large-scale dynamic sparse weights. The main obstacles are:
+1. The `tf.Variable` is fixed size on running time and saved in continuous memory and unable to add or remove weights dynamically.
+2. The typical training mode of full updating in TF is very inefficient when training xTB models.
+3. Usually, the sparse weights are saved in hash table form (K-V pairs for sparse feature ID and sparse weights) but the optimizers of TF cann't train directly sparse weights and accumulators(slots) saved in hash tables.
+
+For the above reasons, the official version of TensorFlow can only be used for offline model hyperparameter tuning with fixed size features, and the model can not exceed 100GB generally. The application in the industrial scene is very limited. At present, We customized TensorFlow and successfully applied it to large-scale sparse features training of industrial recommendation scene. We hope to contribute this work of general components to the community.
+
+## Design Proposal
+
+### The ideal semantics:
+
+```python
+import tensorflow.dynamic_embedding as dye
+
+# sparse weights defination:
+w = dye.get_variable(name="dynamic_embeddings",
+                     devices=["/job:ps/replica:0/task:0/CPU:0",],
+                     dim=8)
+z = dye.embedding_lookup(params=w,
+                         ids=x, # x is got from samples.
+                         name="wide-sparse-weights")
+                         # z should be a trainable and variable-size instance.
+# forward graph
+
+pred = tf.XXX(...)
+
+# backward graph
+loss = tf.XXX(pred, label, ...)
+opt = tf.train.AdamOptimizer(learning_rate=0.1)
+update_op = opt.minimize(loss)
+```
+
+## Sparse Domain Isolation
+
+### Overview of Design
```

Comment:
What does the gradient look like? Do you need to define custom gradient function for `embedding_lookup`?

Reply:
> What does the gradient look like? Do you need to define custom gradient function for `embedding_lookup`?

No need to define custom gradient function.
I believe the back-propagation should end after  the updating to TrainableWrap.